pkg/storage/url/urldownload/storage.go

```diff
@@ -0,0 +1,123 @@
+package urldownload
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"path/filepath"
+
+	"github.com/filecoin-project/bacalhau/pkg/config"
+	"github.com/filecoin-project/bacalhau/pkg/storage"
+	"github.com/filecoin-project/bacalhau/pkg/system"
+	"github.com/go-resty/resty/v2"
+	"github.com/rs/zerolog/log"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// a storage driver runs the downloads content
+// from a public URL source and copies it to
+// a local directory in preparation for
+// a job to run - it will remove the folder/file once complete
+
+type StorageProvider struct {
+	LocalDir   string
+	HTTPClient *resty.Client
+}
+
+func NewStorageProvider(cm *system.CleanupManager) (*StorageProvider, error) {
+	// TODO: consolidate the various config inputs into one package otherwise they are scattered across the codebase
+	dir, err := ioutil.TempDir(config.GetStoragePath(), "bacalhau-url")
+	if err != nil {
+		return nil, err
+	}
+
+	client := resty.New()
+	// Setting output directory path, If directory not exists then resty creates one
+	client.SetOutputDirectory(dir)
+
+	storageHandler := &StorageProvider{
+		HTTPClient: client,
+		LocalDir:   dir,
+	}
+
+	log.Debug().Msgf("URL download driver created with output dir: %s", dir)
+	return storageHandler, nil
+}
+
+func (sp *StorageProvider) IsInstalled(ctx context.Context) (bool, error) {
+	_, span := newSpan(ctx, "IsInstalled")
+	defer span.End()
+	return true, nil
+}
+
+func (sp *StorageProvider) HasStorageLocally(ctx context.Context, volume storage.StorageSpec) (bool, error) {
+	_, span := newSpan(ctx, "HasStorageLocally")
+	defer span.End()
+	return false, nil
+}
+
+// Could do a HEAD request and check Content-Length, but in some cases that's not guaranteed to be the real end file size
+func (sp *StorageProvider) GetVolumeSize(ctx context.Context, volume storage.StorageSpec) (uint64, error) {
+	return 0, nil
+}
+
+func (sp *StorageProvider) PrepareStorage(ctx context.Context, storageSpec storage.StorageSpec) (storage.StorageVolume, error) {
+	_, span := newSpan(ctx, "PrepareStorage")
+	defer span.End()
+
+	_, err := IsURLSupported(storageSpec.URL)
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+
+	outputPath, err := ioutil.TempDir(sp.LocalDir, "*")
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+
+	sp.HTTPClient.SetTimeout(config.GetDownloadURLRequestTimeout())
+	_, err = sp.HTTPClient.R().
+		SetOutput(outputPath + "/file").
+		Get(storageSpec.URL)
+	if err != nil {
+		return storage.StorageVolume{}, err
```

Comment:
should we do retries? maybe make a follow-up issue for it.

also: do we currently download all URLs in parallel, because they are separate storage driver instances? this is probably ok for now, but if user provides millions of URLs this might overload the system and we'd want to put them in some sort of capped parallelism queue

Reply:
Good points. Noted https://github.com/filecoin-project/bacalhau/issues/411